### ISSUE TITLE
TypeError: linear_forward(): incompatible function arguments

### ISSUE THREAD

**Issue by kamanphoebe** [laekov/fastmoe#147]
Hi,

I have built a model containing `FMoETransformerMLP()` modules. However, during the forward pass, the following error shows up:
`TypeError: linear_forward(): incompatible function arguments. The following argument types are supported:
    1. (arg0: at::Tensor, arg1: at::Tensor, arg2: at::Tensor, arg3: Optional[at::Tensor]) -> at::Tensor`

Printing out the class of four parameters of `fmoe_cuda.linear_forward()`, namely `global_input_buf`, `fwd_expert_count`, `weights` and `bias`, gives:
`<class 'torch.Tensor'> <class 'torch.Tensor'> <class 'torch.nn.parameter.Parameter'> <class 'torch.nn.parameter.Parameter'>`
It seems that the restriction for types is too strict since `nn.parameter` should be acceptable. Is there any way to solve this error?

**Environment**
- CUDA version: [11.4] and [11.7]
I have tried on two machines with different CUDA versions but got the same error.
- NCCL version: [2.18.1-1+cuda12.1]
I got 2.14.3 by running `torch.cuda.nccl.version()`. Yet the NCCL package always automatically upgrades to 2.18.1 for installation even though I had specified the desired version. Not sure whether it would affect.

**Backtrace**
```
File "/mnt/share/agiuser/luojiawen/moe-incubation/fastmoe/fmoe/transformer.py", line 65, in forward                                              
    output = super().forward(inp)                                                                                                                  
  File "/mnt/share/agiuser/luojiawen/moe-incubation/fastmoe/fmoe/layers.py", line 241, in forward                                                  
    fwd = _fmoe_general_global_forward(                                                                                                            
  File "/mnt/share/agiuser/luojiawen/moe-incubation/fastmoe/fmoe/layers.py", line 61, in _fmoe_general_global_forward                              
    x = expert_fn(x, fwd_expert_count)                                                                                                             
  File "/mnt/share/agiuser/luojiawen/moe-incubation/fastmoe/fmoe/layers.py", line 168, in expert_fn                                                
    outputs.append(self.experts[i](inp_slice, torch.tensor([fwd_expert_count[i]])))                     
  File "/mnt/share/agiuser/luojiawen/miniconda3/envs/moe/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1194, in _call_impl   
    return forward_call(*input, **kwargs)                                                                                                          
  File "/mnt/share/agiuser/luojiawen/moe-incubation/fastmoe/fmoe/transformer.py", line 30, in forward                                              
    x = self.h4toh(x, fwd_expert_count)                                                                                                            
  File "/mnt/share/agiuser/luojiawen/miniconda3/envs/moe/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1194, in _call_impl         
    return forward_call(*input, **kwargs)
  File "/mnt/share/agiuser/luojiawen/moe-incubation/fastmoe/fmoe/linear.py", line 73, in forward
    x = MOELinear.apply(inp, fwd_expert_count, self.weight, self.bias)
  File "/mnt/share/agiuser/luojiawen/moe-incubation/fastmoe/fmoe/linear.py", line 19, in forward
    global_output_buf = fmoe_cuda.linear_forward(
TypeError: linear_forward(): incompatible function arguments. The following argument types are supported:
    1. (arg0: at::Tensor, arg1: at::Tensor, arg2: at::Tensor, arg3: Optional[at::Tensor]) -> at::Tensor
 ```

**Comment by laekov** [comment]
That is unexpected. PyTorch should do the conversion automatically. 

As a hotfix, you can change `weight` and `bias` to `weight.detach()` and `bias.detach()` at `fastmoe/fmoe/linear.py:20` (and accordingly at l.30).

If you would like to help us dig deeper into this issue, can you please provide your version of PyTorch (and how it is installed) and PyBind 11? It would be better if you can provide us with a minimum script that can reporduce this issue.

**Comment by kamanphoebe** [comment]
Thanks for the quick reply!

My PyTorch and PyBind11 version are stated below. Weirdly, adding `.detach()` didn't help. Even replacing `weight` and `bias` with `torch.ones_like(weight)` and `torch.ones_like(bias)` has no effect. It seems that the error appears regardless of the types. Instead, everything went well if not using _any_ activation function for the expert computation, i.e., commenting out `fastmoe/fmoe/transformer.py:29`. 

I also wrote a simple script [debug.py](https://github.com/laekov/fastmoe/files/11407817/debug.zip) for reproducing the issue. There are two model classes defined in the script which contain a `FMoETransformerMLP()` module. However, one of them can successfully finish the forward pass while the other fails given the same input. 

Hope it helps :)

**Environment**
- PyTorch version: [1.13.1]
It is installed in a conda environment via the command 
`conda install pytorch torchvision torchaudio pytorch-cuda=11.7 -c pytorch -c nvidia`.
- PyBind11: [2.10.4]
I installed PyBind11 after seeing your comment and the error remains.
- CUDA version: [11.4] and [11.7]
I have tried on two machines with different CUDA versions but got the same error.
- NCCL version: [2.18.1-1+cuda12.1]
I got 2.14.3 by running torch.cuda.nccl.version(). Yet the NCCL package always automatically upgrades to 2.18.1 for installation even though I had specified the desired version. Not sure whether it would affect.
